File: 124/my_solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_overall = -10**10
        def maxPath(root) -> list:
            nonlocal max_overall
            if root == None: return [0,0,0]
            if root.left == None and root.right == None:
                if root.val > max_overall: max_overall = root.val
                return [root.val,root.val,root.val]

            L = maxPath(root.left)
            R = maxPath(root.right)
            X = max(L[1],L[2]) if max(L[1],L[2]) > 0 else 0
            Y = max(R[1],R[2]) if max(R[1],R[2]) > 0 else 0
            A = root.val + X + Y
            B = root.val + X
            C = root.val + Y
            
            max_now = max([root.val,A,B,C])

            if max_now > max_overall:
                max_overall = max_now
            
            return [A,B,C]

        maxPath(root)
        return max_overall

File: 222/my_solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int depth;
    bool leafBeforeDepth = false;
public:
    int getDepth(TreeNode *root){
        if (root == nullptr) return 0;
        return 1 + getDepth(root->left);
    }

    int countLeavesAtDepth(TreeNode *root, int level){
        if (root == nullptr) return 0;
        if (root->left == nullptr && root->right == nullptr){
            if (level == depth) return 1;
            else {
                leafBeforeDepth = true;
                return 0;
            }
        }
        else if (root->left != nullptr && root->right == nullptr){
            leafBeforeDepth = true;
            return 1;
        }

        int count = 0;
        if (leafBeforeDepth == false) count += countLeavesAtDepth(root->left, level + 1);
        if (leafBeforeDepth == false) count += countLeavesAtDepth(root->right, level + 1);

        return count;
    }

    int countNodes(TreeNode* root) {
        if (root == nullptr) return 0;

        depth = getDepth(root);
        cout << depth;
        int leaves = countLeavesAtDepth(root, 1);

        return pow(2,depth-1) + leaves - 1;
    }
};